### ISSUE TITLE
Error with Models using both float and long as input

### ISSUE THREAD

**Issue by harshraj22** [pytorch/captum#836]
## ❓ Question
I was trying captum on [tab transformer](https://github.com/harshraj22/DeepLearning/tree/main/implementations/TabTransformer), trying to find which features affect the result by using `integrated_graditents`. As seen in the [picture](https://user-images.githubusercontent.com/46635452/145711600-22e3ccf9-f45a-49b6-b029-2315e8767b80.png) of the readme file, it accepts two inputs:
- Categorical features (which is of type `torch.long` or similar)
- Continious features (which is of type `torch.float`)

Inside the model, the categorical features are passed through the embedding layer before further operations, converting into `torch.float`.
When running `integrated_gradients.attribute`, the categorical features are converted into `torch.float` before passing into `forward` method, giving error.

<strong>Note:</strong> Please note that this is different from a text classification model (where whole input is `torch.long`) or image classification model (where whole input is `torch.float`).

Attached below is screenshot of error, where I debug the passed tensor by printing it before the [forward method of transformer](https://github.com/harshraj22/DeepLearning/blob/d67edc978f2543d541e124124795f78a81c4fbed/implementations/TabTransformer/models/transformer.py#L56)
```python
    def forward(self, x_cat, x_cont):
        batch_size, num_cat_cols = x_cat.shape
        
        print(f'[transformer.py]: x_cat: {x_cat.shape} {x_cat}')
        # x_cat.shape: (num_cat_cols, embed_dim)
        x_cat = self.embed(x_cat)
```
And this is the contents of `rough.py` used to run integrated graditents calculation (`python3 rough.py`) 
```python3
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, Dataset
from torch.optim import Adam

from tqdm import tqdm
from pprint import pprint
import wandb
import hydra
import logging
import pathlib

from models.transformer import TabTransformer
from data_loader.datasets import BlastcharDataset
from utils.utils import Phase

from captum.attr import IntegratedGradients
from captum.attr import GradientShap
from captum.attr import Occlusion
from captum.attr import NoiseTunnel
from captum.attr import visualization as viz

logging.basicConfig(level=logging.ERROR)

@hydra.main(config_path="conf", config_name="config")
def main(cfg):
    # wandb.config.update(dict(cfg))
    # Initialize the dataset
    blastchar_dataset = BlastcharDataset(cfg.dataset.path)
    NUM_CATEGORICAL_COLS = blastchar_dataset.num_categorical_cols
    NUM_CONTINIOUS_COLS = blastchar_dataset.num_continious_cols
    EMBED_DIM = 32

    # initialize the model with its arguments
    mlp = nn.Sequential(
            nn.Linear(NUM_CATEGORICAL_COLS * EMBED_DIM + NUM_CONTINIOUS_COLS, 50),
            nn.ReLU(),
            nn.BatchNorm1d(50),
            nn.Dropout(cfg.params.dropout),

            nn.Linear(50, 20),
            nn.ReLU(),
            nn.BatchNorm1d(20),
            nn.Dropout(cfg.params.dropout),

            nn.Linear(20, blastchar_dataset.num_classes)
    )

    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    model = TabTransformer(blastchar_dataset.num_categories, mlp, embed_dim=EMBED_DIM).to(device)
    # model.load_state_dict(torch.load('/nfs_home/janhavi2021/spring2021/tabTransformer/saved/models/400epochs.pth'))

    pred_label_idx = 0
    # input = blastchar_dataset[0].squeeze(0)
    cat, cont, _ = blastchar_dataset[0]
    cat, cont = cat.unsqueeze(0).long(), cont.unsqueeze(0).long()
    # input = torch.cat((cat, cont), dim=0)
    input = (cat, cont)

    print(f' \n\n\n[rough.py]: {cat.shape}{cont.shape}')
    print(f'[rough.py]: {cat}')
    integrated_gradients = IntegratedGradients(model)
    attributions_ig = integrated_gradients.attribute(input, target=pred_label_idx, n_steps=200)
    print(attributions_ig)

main()
```

![image](https://user-images.githubusercontent.com/46635452/148019672-fa9f03b6-c497-4a52-bd7b-fe43ccf1f757.png)

## To Reproduce

1. Set up the tab transformers repo (either setup Balstchar dataset or generate the inputs of required shape and datatype: x_cat: 16 integers converted to `torch.long`, x_cont: 4 floats converted to `torch.float` as shown by first print line in the screenshot)
2. Try to run the integrated gradients code (the contents of which is shown in `rough.py` given above)

## Expected behavior
It should give the attributes of shape 20 (16 + 4) describing which feature (from the categorical and continious features) affects the result most.

## Environment
Describe the environment used for Captum

 - Captum / PyTorch Version (e.g., 1.0 / 0.4.0): captum==0.4.1, torch '1.9.0+cu102'
 - OS (e.g., Linux): linux
 - How you installed Captum / PyTorch (`conda`, `pip`, source): pip install captum
 - Build command you used (if compiling from source):
 - Python version: Python 3.8.8 
 - CUDA/cuDNN version:
 - GPU models and configuration: N/A
 - Any other relevant information:

## Additional context
I went through the codebase, and it seems [this line](https://github.com/pytorch/captum/blob/38b57082d22854013c0a0b80a51c0b85269afdaf/captum/attr/_core/integrated_gradients.py#L329) is converting the categorical features to `torch.float` (alpha being a float). Let me know if there is any alternate way to calculate the attributions for this case using captum.

**Comment by harshraj22** [comment]
Update: I also tried LayerIntegratedGradient as described in [Model Interpretability for multimodal models](https://captum.ai/tutorials/Multimodal_VQA_Interpret)

```python
# rough.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader, Dataset
from torch.optim import Adam

from tqdm import tqdm
from pprint import pprint
import wandb
import hydra
import logging
import pathlib

from models.transformer import TabTransformer
from data_loader.datasets import BlastcharDataset
from utils.utils import Phase

from captum.attr import IntegratedGradients, LayerIntegratedGradients
from captum.attr import GradientShap
from captum.attr import (
    IntegratedGradients,
    LayerIntegratedGradients,
    TokenReferenceBase,
    configure_interpretable_embedding_layer,
    remove_interpretable_embedding_layer,
    visualization
)
from captum.attr import NoiseTunnel
from captum.attr import visualization as viz
from captum.attr._utils.input_layer_wrapper import ModelInputWrapper

logging.basicConfig(level=logging.ERROR)

@hydra.main(config_path="conf", config_name="config")
def main(cfg):
    # wandb.config.update(dict(cfg))
    # Initialize the dataset
    blastchar_dataset = BlastcharDataset(cfg.dataset.path)
    NUM_CATEGORICAL_COLS = blastchar_dataset.num_categorical_cols
    NUM_CONTINIOUS_COLS = blastchar_dataset.num_continious_cols
    EMBED_DIM = 32

    # initialize the model with its arguments
    mlp = nn.Sequential(
            nn.Linear(NUM_CATEGORICAL_COLS * EMBED_DIM + NUM_CONTINIOUS_COLS, 50),
            nn.ReLU(),
            nn.BatchNorm1d(50),
            nn.Dropout(cfg.params.dropout),

            nn.Linear(50, 20),
            nn.ReLU(),
            nn.BatchNorm1d(20),
            nn.Dropout(cfg.params.dropout),

            nn.Linear(20, blastchar_dataset.num_classes)
    )

    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    model = TabTransformer(blastchar_dataset.num_categories, mlp, embed_dim=EMBED_DIM).to(device)
    # model.load_state_dict(torch.load('/nfs_home/janhavi2021/spring2021/tabTransformer/saved/models/400epochs.pth'))
    print(model)
    model = ModelInputWrapper(model)

    pred_label_idx = 0
    # input = blastchar_dataset[0].squeeze(0)
    cat, cont, _ = blastchar_dataset[0]
    cat, cont = cat.unsqueeze(0).long(), cont.unsqueeze(0).long()
    # input = torch.cat((cat, cont), dim=0)
    input = (cat, cont)

    print(f' \n\n\n[rough.py]: {cat.shape}{cont.shape}')
    print(f'[rough.py]: {cat}')


    interpretable_embedding = configure_interpretable_embedding_layer(model, 'module.embed')
    token_reference = TokenReferenceBase(reference_token_idx=1)
    attr = LayerIntegratedGradients(model, [model.module.embed], model.module.embed)

    attributions = attr.attribute(inputs=(cat, cont), baselines=(torch.zeros_like(cat, dtype=torch.long), torch.zeros_like(cont, dtype=torch.float32)), target=1, n_steps=30)

    # lig = LayerIntegratedGradients(model, model.embed)
    # attributions_ig, delta = lig.attribute((cat, cont), n_steps=500, return_convergence_delta=True, baselines=(torch.zeros_like(cat, dtype=torch.long), torch.zeros_like(cont, dtype=torch.float32)))
    # integrated_gradients = IntegratedGradients(model)
    # attributions_ig = integrated_gradients.attribute(input, target=pred_label_idx, n_steps=200)
    # print(attributions_ig)

main()
```
![image](https://user-images.githubusercontent.com/46635452/148031253-f20f8421-54a6-4359-91bf-2d34e5f449bb.png)

TBH, the tutorials linked on [website](https://captum.ai/tutorials/Multimodal_VQA_Interpret) are not very clear.

**Comment by bilalsal** [comment]
Hi @harshraj22 ,

I am very happy you figured out an excellent solution to get the attributions for the categorical features.

In your 1st post you also correctly identified the challenge of using `IntegratedGradients` to handle a mixed scalar + categorical input.
Some solutions:

- Use other attribution algorithms that do NOT interpolate between the input and the baseline. `IntegratedGradients` relies on such interpolation. You can try using simple perturbation-based algorithms such as `FeatureAblation` and `Occlusion` or simplified gradient-based algorithms such as `GuidedBackprop` and `LRP`.
- Use `IntegratedGradients` twice: Once for the 16 categorical features as you did, and once for the scalar features. You can try applying `requires_grad=False` to the `cat` tensor to exclude it from the attribution, or define a wrapper that takes only the continuous features as an input and feeds the categorical features subsequently. That way, Captum won't perform interpolation with the baseline that caused the error as you identified.

Hope this helps

**Comment by harshraj22** [comment]
@bilalsal Thanks for your comments and suggestions.
I am new to integrated gradients, and I am a little confused about your suggestion 2.

As you pointed out, I use integrated gradients twice (as depicted in tutorial for Visual Question Answering)
```python

    outs = model(*input)
    preds = outs.argmax(-1)


    attr = LayerIntegratedGradients(model, [model.module.embed, model.module.layer_norm])

    attributions, _ = attr.attribute(inputs=(cat, cont), baselines=(torch.zeros_like(cat, dtype=torch.long), torch.zeros_like(cont, dtype=torch.float32)), target=preds.detach(), n_steps=30, return_convergence_delta=True)

    print(f'attributions: {attributions[0].shape, attributions[1].shape}')
    pprint(torch.cat((attributions[0].sum(dim=2), attributions[1]), dim=1))

```

However, it returns tuple of attributions (16 and 4) as different. As far as I understand, the effect of features across these 16 can be compared by their attributions value, ie. if attributions[cat_1] > attributions[cat_2], then feature 1 impacts the decision than feature 2. But can we also compare the features across the inputs ?
I mean, can I say, if attributions[cat_1] > attributions[cont_1], then cat_1 impacts more than cont_1 ?

**Comment by bilalsal** [comment]
Hi @harshraj22 ,

your solution to pass multiple layers is the same `.attribute()` call is even better.
Now I understand your concern about the comparability of the computed attributions across modality.
If you sum up the absolute attributions for all 16 categorical features, the sum will be comparable to that of all 4 scalar features, in the sense that you can estimate which modality has contributed more to the outcome.
But I agree it might be less informative to compare the attributions of two individual features having different modality.

Hope this helps

**Comment by harshraj22** [comment]
Thanks for the answers.
Closing issue as found the answer.
